pkg/registry/node/etcd/etcd.go

```diff
@@ -90,5 +91,39 @@ var _ = rest.Redirector(&REST{})
 
 // ResourceLocation returns a URL to which one can send traffic for the specified node.
 func (r *REST) ResourceLocation(ctx api.Context, id string) (*url.URL, http.RoundTripper, error) {
-	return node.ResourceLocation(r, r.connection, r.proxyTransport, ctx, id)
+	return node.ResourceLocation(r, r, r.proxyTransport, ctx, id)
+}
+
+var _ = client.ConnectionInfoGetter(&REST{})
+
+func (r *REST) getKubeletPort(host string, ctx api.Context) (int, error) {
+	// We probably shouldn't care about context when looking for Node object.
+	obj, err := r.Get(ctx, host)
+	if err != nil {
+		return 0, err
+	}
+	node := obj.(*api.Node)
+	if node == nil {
+		return 0, fmt.Errorf("Unexpected object type: %#v", node)
+	}
+	if node.Status.DaemonEndpoints.KubeletEndpoint.Port == 0 {
+		return -1, nil
+	}
+	return node.Status.DaemonEndpoints.KubeletEndpoint.Port, nil
+}
+
+func (c *REST) GetConnectionInfo(ctx api.Context, host string) (string, uint, http.RoundTripper, error) {
+	scheme, port, transport, err := c.connection.GetConnectionInfo(ctx, host)
+	if err != nil {
+		return "", 0, nil, err
+	}
+	daemonPort, err := c.getKubeletPort(host, ctx)
+	if err != nil {
+		return "", 0, nil, err
+	}
+	if daemonPort >= 0 {
```

Comment:
I'd expect `if daemonPort > 0`... and `getKubeletPort` could just blindly `return node.Status.DaemonEndpoints.KubeletEndpoint.Port, nil`